test/tests_for_integration/auth_controller.py

```diff
@@ -0,0 +1,160 @@
+    """
+
+    def __init__(
+        self,
+        jars_dir: Path,
+        mongo_host: str,
+        mongo_db: str,
+        root_temp_dir: Path,
+        mongo_user: str = None,
+        mongo_pwd: str = None,
+    ):
+        """
+        Create and start a new Auth service. An unused port will be selected for the server.
+
+        :param jars_dir: The path to the lib/jars dir of the KBase Jars repo
+            (https://github.com/kbase/jars), e.g /path_to_repo/lib/jars.
+        :param mongo_host: The address of the MongoDB server to use as the Auth service database,
+            e.g. localhost:27017.
+        :param mongo_db: The database in which to store Auth data.
+        :param root_temp_dir: A temporary directory in which to store Auth data and log files.
+            The files will be stored inside a child directory that is unique per invocation.
+        :param mongo_user: if the MongoDB server requires authentication, the user name.
+        :param mongo_pwd: if the MongoDB server requires authentication, the user password.
+        """
+        if not jars_dir or not os.access(jars_dir, os.X_OK):
+            raise TestException(
+                "jars_dir {} does not exist or is not executable.".format(jars_dir)
+            )
+        if not mongo_host:
+            raise TestException("mongo_host must be provided")
+        if not mongo_db:
+            raise TestException("mongo_db must be provided")
+        if not root_temp_dir:
+            raise TestException("root_temp_dir is None")
+
+        if bool(mongo_user) ^ bool(mongo_pwd):  # xor
+            raise TestException(
+                "Neither or both of mongo_user and mongo_pwd is required"
+            )
+
+        jars_dir = jars_dir.resolve()
+        class_path = self._get_class_path(jars_dir)
+
+        # make temp dirs
+        root_temp_dir = root_temp_dir.absolute()
+        os.makedirs(root_temp_dir, exist_ok=True)
+        self.temp_dir = Path(
+            tempfile.mkdtemp(prefix="AuthController-", dir=str(root_temp_dir))
+        )
+
+        self.port = test_utils.find_free_port()
+
+        template_dir = self.temp_dir.joinpath("templates")
+        self._install_templates(jars_dir, template_dir)
+
+        command = [
+            "java",
+            "-classpath",
+            class_path,
+            "-DAUTH2_TEST_MONGOHOST=" + mongo_host,
+            "-DAUTH2_TEST_MONGODB=" + mongo_db,
+            "-DAUTH2_TEST_TEMPLATE_DIR=" + str(template_dir),
+            _AUTH_CLASS,
+            str(self.port),
+        ]
+        if mongo_user:
+            command.insert(5, "-DAUTH2_TEST_MONGOPWD=" + mongo_pwd)
+            command.insert(5, "-DAUTH2_TEST_MONGOUSER=" + mongo_user)
+
+        self._outfile = open(self.temp_dir.joinpath("auth.log"), "w")
+
+        self._proc = subprocess.Popen(
+            command, stdout=self._outfile, stderr=subprocess.STDOUT
+        )
+
+        for count in range(40):
```

Comment:
Could make `40` a variable initialised at the top of the file so that it's easy to alter if necessary.

Reply:
done